src/hamiltonian.jl

```diff
@@ -41,12 +41,16 @@ function ∂H∂θ(h::Hamiltonian, θ::AbstractVecOrMat)
     return DualValue(res[1], -res[2])
 end
 
-∂H∂r(h::Hamiltonian{<:UnitEuclideanMetric,<:GaussianKinetic}, r::AbstractVecOrMat) = copy(r)
-function ∂H∂r(h::Hamiltonian{<:DiagEuclideanMetric,<:GaussianKinetic}, r::AbstractVecOrMat)
-    return h.metric.M⁻¹ .* r
+∂H∂r(h::Hamiltonian{<:UnitEuclideanMetric,<:GaussianKinetic}, r::AbstractArray) = copy(r)
+function ∂H∂r(h::Hamiltonian{<:DiagEuclideanMetric,<:GaussianKinetic}, r::AbstractArray)
+    out = similar(r)
+    out .= h.metric.M⁻¹ .* r
+    return out
 end
-function ∂H∂r(h::Hamiltonian{<:DenseEuclideanMetric,<:GaussianKinetic}, r::AbstractVecOrMat)
-    return h.metric.M⁻¹ * r
+function ∂H∂r(h::Hamiltonian{<:DenseEuclideanMetric,<:GaussianKinetic}, r::AbstractArray)
+    out = similar(r)
+    mul!(out, h.metric.M⁻¹, r)
```

Comment:
Generally this is not sufficient, you need something like
```suggestion
    (; M⁻¹) = h.metric
    out = similar(r, promote_type(eltype(r),eltype(M⁻¹)))
    mul!(out, M⁻¹, r)
```

But the question is still: Why does the previous code - which in principle is much nicer since it doesn't require type heuristics - not work? Is it an upstream bug in ComponentArrays? Or because the hamiltonian was set up incorrectly?

Reply:
I believe the main issue here is that when we try to compute the product `Matrix{Float64}*ComponentVector{Float64}`, we would have `Vector{Float64}`, so I think maybe it's a ComponentArrays.jl thing that missing dispatches for such linear algebra operations?

```julia
julia> x = ComponentArray(a=5, b=[(a=20., b=0), (a=33., b=0), (a=44., b=3)], c=c)
ComponentVector{Float64}(a = 5.0, b = ComponentVector{Float64, SubArray{Float64, 1, Vector{Float64}, Tuple{UnitRange{Int64}}, true}, Tuple{Axis{(a = 1, b = 2)}}}[(a = 20.0, b = 0.0), (a = 33.0, b = 0.0), (a = 44.0, b = 3.0)], c = (a = 2.0, b = [1.0, 2.0]))

julia> rand(10,10)*x
10-element Vector{Float64}:
 53.280246927016925
 57.22007605159049
 49.206319389153876
 63.135902809337765
 50.88459580133911
 73.58796576720756
 46.62055260143546
 44.463608176249025
 68.52373868783468
 46.42311997871232
```